services/metadata/src/metadata/services/avus.clj

```diff
@@ -141,6 +141,56 @@
                             (:avus (when template-id
                                          (update-metadata-template-avus
                                            user-id data-id data-type template-id metadata :publish-update false)))))]
-      (persistence/remove-orphaned-avus data-id avu-ids)))
+      (persistence/remove-orphaned-avus data-type data-id avu-ids)))
   (amqp/publish-metadata-update user-id data-id)
   (list-metadata-template-avus data-id))
+
+(defn filter-targets-by-attr-value
+  "Filters the given target IDs by returning a list of any that have the given attr and value applied."
+  [attr value target-types target-ids]
+  {:target-ids (map :target_id
+                    (persistence/filter-targets-by-attr-values target-types target-ids attr [value]))})
+
+(defn- format-avu
+  "Formats the given AVU for adding or updating."
+  [target-type target-id {:keys [attr] :as avu}]
+  (-> (select-keys avu [:id :value :unit])
+      (assoc
+        :attribute   attr
+        :target_type target-type
+        :target_id   target-id)))
+
+(defn list-avus
+  [target-type target-id]
+  {:avus (persistence/avu-list target-type target-id)})
+
+(defn update-avus
+  [user-id target-type target-id {avus :avus}]
+  (transaction
+   (doseq [avu avus]
+     (persistence/add-or-update-avu user-id
+                                    (format-avu target-type target-id avu))))
+  (amqp/publish-metadata-update user-id target-id)
+  (list-avus target-type target-id))
+
+(defn- remove-orphaned-avus
+  "Removes any AVU for the given target-type and target-id that does not have a matching ID in the given
+   set of avus."
+  [target-type target-id avus]
+  (->> avus
+       (map :id)
+       (remove nil?)
+       (persistence/get-avus-by-ids)
+       (map :id)
+       (persistence/remove-orphaned-avus target-type target-id)))
+
+(defn set-avus
+  "Sets AVUs for the given user's data item."
+  [user-id target-type target-id {avus :avus}]
+  (transaction
+   (remove-orphaned-avus target-type target-id avus)
+   (doseq [avu avus]
+     (persistence/add-or-update-avu user-id
+                                    (format-avu target-type target-id avu))))
```

Comment:
The new `set-avus` metadata endpoint will add/update AVUs differently than the current `POST /filesystem/data/{data-id}/avus` endpoint, since it's now a requirement that we allow multiple AVUs on a target to share the same attribute with different values (this is the only way to categorize an app in multiple categories with AVUs, but it's also been requested for data AVUs).

The current data "Set AVUs" endpoint attempts to update an AVU based on either its ID or its attribute name. The new `set-avus` endpoint instead assumes that only AVUs with IDs in the request body require an update, and AVUs without IDs should be added as new.

So this new `set-avus` endpoint first deletes all AVUs on the target that do not have a matching ID in the request, before updating or adding new AVUs. This will help prevent "unique constraint" postgres exceptions in the case where the client deletes an AVU, then updates an existing AVU with the same values in the deleted AVU, then submits this final state in 1 `set` request.

For example, suppose a target has the following AVUs:
```clojure
[{:id 1, :attr "a", :value "b", :unit "c"}
 {:id 2, :attr "d", :value "e", :unit "f"}]
```
Then the user decides to delete the AVU with `:id 2`, and before submitting a request to save the AVUs, updates the AVU with `:id 1` to `{:id 1 :attr "d" :value "e" :unit "f"}` in the client. If this function did not delete AVUs whose IDs were not included in the request first, then the AVU with `:id 2` would cause postgres to throw a "unique constraint" exception when it attempted to update the AVU with `:id 1` to the same values.